octopusdeploy_framework/resource_machine_proxy.go

```diff
@@ -0,0 +1,143 @@
+package octopusdeploy_framework
+
+import (
+	"context"
+	"github.com/OctopusDeploy/go-octopusdeploy/v2/pkg/core"
+	"github.com/OctopusDeploy/go-octopusdeploy/v2/pkg/proxies"
+	"github.com/OctopusDeploy/terraform-provider-octopusdeploy/internal/errors"
+	"github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy_framework/schemas"
+	"github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy_framework/util"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+var _ resource.Resource = &machineProxyResource{}
+
+type machineProxyResource struct {
+	*Config
+}
+
+func NewMachineProxyResource() resource.Resource {
+	return &machineProxyResource{}
+}
+
+func (r *machineProxyResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = util.GetTypeName(schemas.MachineProxyResourceName)
+}
+
+func (r *machineProxyResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schemas.MachineProxySchema{}.GetResourceSchema()
+}
+
+func (r *machineProxyResource) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	r.Config = ResourceConfiguration(req, resp)
+}
+
+func (r *machineProxyResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var plan schemas.MachineProxyResourceModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	machineProxy := mapMachineProxyModelToRequest(&plan)
+	createdProxy, err := proxies.Add(r.Client, machineProxy)
+	if err != nil {
+		resp.Diagnostics.AddError("Error creating machine proxy", err.Error())
+		return
+	}
+
+	proxyModel := mapMachineProxyRequestToModel(createdProxy, &plan)
+
+	diags := resp.State.Set(ctx, proxyModel)
+	resp.Diagnostics.Append(diags...)
+}
+
+func (r *machineProxyResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var state schemas.MachineProxyResourceModel
+	diags := req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	machineProxy, err := proxies.GetByID(r.Client, state.SpaceID.ValueString(), state.ID.ValueString())
```

Comment:
I'm sure it's handled, but just for my clarity: if the user doesn't provide a SpaceId on the Datasource, will this `state.SpaceID` be automatically set to the default Space from the Provider config?

Reply:
yes, that happens inside the Go client, if the space ID is "" (go will not let it be nil) we will use the space provided on the provider, if one is not provided by the provider it will use default space, if you don't have a default space set it will  error gracefully.